File: ClientLibrary/robot/Tank.cs
using System;
using System.Threading.Tasks;
using BaseLibrary;
using BaseLibrary.command;
using BaseLibrary.command.common;
using BaseLibrary.command.handshake;
using BaseLibrary.command.tank;
using BaseLibrary.equip;
using BaseLibrary.protocol;

namespace ClientLibrary.robot {

    /// <summary>
    /// Instances represent robot who can shoot bullets.
    /// </summary>
    public class Tank : ClientRobot {

        /// <summary>
        /// Gun witch robot has.
        /// </summary>
        public Gun GUN { get; private set; }

        public Tank(String name) : base(name) { }

        public Tank(String name, String teamName) : base(name, teamName) {}

        /// <summary>
        /// Shoot bullet.
        /// </summary>
        /// <param name="angle">in degree. 0 = 3 hour. 90 = 6 hour and so on.</param>
        /// <param name="range">how far this robot wants to shot</param>
        public ShootAnswerCommand Shoot(double angle, double range) {
            ShootAnswerCommand answer = new ShootAnswerCommand();
            addRobotTask(ShootAsync(answer, angle, range));
            return answer;
        }

        /// <summary>
        /// Shoot bullet. And sent it to server asynchronously.
        /// </summary>
        /// <param name="destination">Where to fill answer data.</param>
        /// <param name="angle">in degree. 0 = 3 hour. 90 = 6 hour and so on.</param>
        /// <param name="range">how far this robot wants to shot</param>
        private async Task<ShootAnswerCommand> ShootAsync(ShootAnswerCommand destination, double angle, double range) {
            await sendCommandAsync(new ShootCommand(range, angle));
            ShootAnswerCommand answer = receiveCommand<ShootAnswerCommand>();
            destination.FillData(answer);
            return answer;
        }

        /// <inheritdoc />
        public override RobotType GetRobotType() {
            return RobotType.TANK;
        }

        /// <inheritdoc />
        protected override void SetClassEquip(int id) {
            GUN = GUNS_BY_ID[id];
        }

        /// <inheritdoc />
        protected override IClassEquipment GetClassEquip() {
            return GUN;
        }
    }
}

File: Java-Client/RobotBattlefield-Client-Java/src/ClientLibrary/robot/Tank.java

package ClientLibrary.robot;

import java.util.concurrent.CompletableFuture;

import BaseLibrary.communication.command.tank.ShootAnswerCommand;
import BaseLibrary.communication.command.tank.ShootCommand;
import BaseLibrary.equipment.Gun;
import BaseLibrary.equipment.IClassEquipment;

/**
 * Instances represent robot who can shoot bullets.
 */
public class Tank extends ClientRobot {

	/**
	 * Gun witch robot has.
	 */
	private Gun gun;

	/**
	 * Gun witch robot has.
	 */
	public Gun getGun() {
		return gun;
	}

	public Tank(String name) {
		super(name);
	}

	public Tank(String name, String robotTeamName) {
		super(name, robotTeamName);
	}

	/**
	 * Shoot bullet.
	 * 
	 * @param angle
	 *            - in degree. 0 = 3 hour. 90 = 6 hour and so on.
	 * @param range
	 *            - how far this robot wants to shot
	 */
	public ShootAnswerCommand shoot(double angle, double range) {
		ShootAnswerCommand answer = new ShootAnswerCommand();
		addRobotTask(shootAsync(answer, angle, range));
		return answer;
	}

	/**
	 * Shoot bullet. And sent it to server asynchronously.
	 * 
	 * @param destination
	 *            - Where to fill answer data.
	 * @param angle
	 *            - in degree. 0 = 3 hour. 90 = 6 hour and so on.
	 * @param range
	 *            - how far this robot wants to shot
	 */
	private CompletableFuture<Void> shootAsync(ShootAnswerCommand destination, double angle, double range) {
		return sendCommandAsync(new ShootCommand(range, angle))
				.thenAcceptAsync((trash) -> receiveCommandAsync(ShootAnswerCommand.class)
						.thenAcceptAsync((answer) -> destination.FillData(answer)).join());
	}
	
	

	@Override
	public RobotType getRobotType() {
		return RobotType.TANK;
	}

	@Override
	protected void setClassEquip(int id) {
		gun = GUNS_BY_ID.get(id);
	}

	@Override
	protected IClassEquipment getClassEquip() {
		return gun;
	}
}
